docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+  phase: Running
+  selector: cluster.x-k8s.io/cluster-name=workload,cluster.x-k8s.io/deployment-name=workload-md-0
+  resourceHints:
+    memory: 500mb
+    vCPU: "1"
+    GPU: "1"
+```
+
+If a user wishes to override the resources hints as provided from the machine template, they may
+do so by adding the following annotations to their MachineSets or MachineDeployments:
+
+```
+kind: MachineSet
+metadata:
+  annotations:
+      machine.x-k8s.io/GPU: "0"
+      machine.x-k8s.io/memoryMb: "8192"
+      machine.x-k8s.io/vCPU: "2"
+```
+
+### Security Model
+
+This feature should not impact the security model.
+
+No new permissions are needed by the user for the information they will provide, nor by the
+associated processes that will read it.
+
+The addition of metadata about machine sizes will not short circuit any quotas associated with
+the user for the cloud platform they are operating.
+
+### Risks and Mitigations
+
+One risk for this process is that the user experience is slightly convoluted. A user (or cloud provider)
+will need to create these fields for any machine template which will be used to scale from zero.
+This association might not seem obvious at first as the autoscaler will only see the information
+throught the MachineSets and MachineDeployments.
+
+Creating clear documentation about the flow information, and the action of the cluster autoscaler
+will be the first line of mitigating the confusion around this process. Additionally, adding an
+example in the test Docker provider will help to clarify usage for cloud providers.
+
+## Alternatives
+
+An alternative approach to reconciling the resource hints from the machine template to MachineSets
+and MachineDeployments would be for the cluster autoscaler to directly read the machine template
+through the `infrastructureRef` that exists within those objects. This would require the autoscaler
+to query for the machine templates, as well as the required permissions for those objects.
+
+Another alternative would be for cloud providers to dynamically provide this information at run time.
+This may be an approach to consider in the future, but it would still require a method to expose
+the data at the MachineSet and MachineDeployment level. In effect this might look like an
+automated method of achieving the same action as this proposal.
+
+A much larger alternative would be to create a new custom resource that would act as an autoscaling
+abstraction. This new resource would be accessed by both the cluster autoscaler and the Cluster API
+controllers, as well as potentially another operator to own its lifecycle. This approach would
+provide the cleanest separation between the components, and allow for future features in a contained
+environment. The downside is that this approach requires the most engineering and design work to
+accomplish.
+
+## Upgrade Strategy
+
+As this field is optional, it should not negatively affect upgrades. That said, care should be taken
```

Comment:
Background (silly) question:
Is there a risk of conflicts between a MachineDeployment rolling out new replicas and an autoscaler trying managing the same MachineDeployment, especially now that we are implementing new scale in/scale out algorithms?

Reply:
i'll have to think about this a little more. in general, the autoscaler will not try to do anything until it sees either unscheduled pods or under-utilized nodes, this might be an area where we would want to disable autoscaling on a MachineDeployment if it is re-deploying a set of machines.